### PR TITLE
Master groupby m2m tests boi

### PR DESCRIPTION
**Pull request by brboi**
https://github.com/odoo/enterprise/pull/20477
task-2633123

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/tests/legacy/views/kanban_tests.js [hunk 1]
```diff
@@ -542,6 +588,38 @@ QUnit.module('Views', {
         kanban.destroy();
     });
 
+    QUnit.test('prevent deletion when grouped by many2many field', async function (assert) {
```
Comment (aab-odoo) [inline addons/web/static/tests/legacy/views/kanban_tests.js:591]: this one probably doesn't make sense. This mainly concerns x2manys, in which case the view isn't grouped. Here you mostly test the arch you wrote, and an implementation detail (`widget.deletable`).

File: addons/web/static/tests/legacy/views/kanban_tests.js [hunk 2]
```diff
@@ -3123,6 +3201,50 @@ QUnit.module('Views', {
...
+            'second column should have correct title');
+        assert.strictEqual(kanban.$('.o_kanban_group:first .o_kanban_record').text(), 'yopblip',
+            "first column should have 2 records");
+        assert.strictEqual(kanban.$('.o_kanban_group:last .o_kanban_record').text(), 'yopgnapblip',
+            "second column should have 3 records");
+
+        // drag&drop a record in another column
+        const $record = kanban.$('.o_kanban_group:nth-child(1) .o_kanban_record:first');
+        const $group = kanban.$('.o_kanban_group:nth-child(2)');
+        await testUtils.dom.dragAndDrop($record, $group);
```
Comment (aab-odoo) [inline addons/web/static/tests/legacy/views/kanban_tests.js:3239]: visually, the user can't drag a record outside its column, right?

File: addons/web/static/tests/legacy/views/list_tests.js [hunk 3]
```diff
@@ -1296,6 +1296,203 @@ QUnit.module('Views', {
...
+                </tree>`,
+            viewOptions: { hasActionMenus: true },
+            groupBy: ['m2m'],
+        });
+
+        await testUtils.dom.click(list.$('.o_group_header:first')); // open first group
+        await testUtils.dom.click(list.$('.o_data_row:eq(0) .o_list_record_selector input'));
+        assert.containsOnce(list.el, 'div.o_control_panel .o_cp_action_menus');
+        assert.containsNone(list.el, 'div.o_control_panel .o_cp_action_menus .o_dropdown_menu',
+            "should not have dropdown as delete item is not there");
```
Comment (aab-odoo) [inline addons/web/static/tests/legacy/views/list_tests.js:1409]: reload without the groupby and check it is there (sanity check)

File: addons/web/static/tests/legacy/views/list_tests.js [hunk 4]
```diff
@@ -1296,6 +1296,203 @@ QUnit.module('Views', {
...
+        await testUtils.dom.click(list.$('.o_group_header:first')); // open first group
+        await testUtils.dom.click(list.$('.o_group_header:eq(1)')); // open second group
+        assert.containsOnce(list, 'tbody:eq(1) .o_list_char:contains("yop")',
+            "should have yop record in first group");
+        assert.containsOnce(list, 'tbody:eq(3) .o_list_char:contains("yop")',
+            "should have yop record in second group");
+
+        await testUtils.dom.click(list.$('.o_data_row:eq(0) .o_list_record_selector input'));
+        await testUtils.dom.click(list.$('.o_data_row:eq(0) .o_data_cell:first'));
+        await testUtils.fields.editAndTrigger(list.$('.o_data_row:eq(0) .o_data_cell:first .o_input'), 'xyz', ['input', 'change']);
```
Comment (aab-odoo) [inline addons/web/static/tests/legacy/views/list_tests.js:1438]: you can use editInput here

**Review by aab-odoo — COMMENTED** [review]
I kinda skipped the diff in the mock_servers, as they are tested by the tests themselves.

File: addons/web/static/tests/legacy/views/kanban_tests.js (hunk 1 above)
Comment (brboi) [reply addons/web/static/tests/legacy/views/kanban_tests.js:591]: It is in use.
e.g. here https://github.com/odoo/odoo/blob/3fa3676240b48cce049d79f20292ac1eea8dd727/addons/crm/views/crm_lead_views.xml#L550

**Review by aab-odoo — APPROVED** [review]

File: addons/web/static/tests/helpers/mock_server.js [hunk 5]
```diff
@@ -1578,6 +1649,16 @@ export class MockServer {
                     }
                     criterion = [criterion[0], "in", childIDs];
                 }
+                // In case of many2many field, if domain operator is '=' generally change it to 'in' operator
+                const field = model.fields[criterion[0]] || {};
+                if (field.type === "many2many" && criterion[1] === "=") {
+                    if (criterion[2] === false) {
+                        // if undefined value asked, domain.js require equality with empty array
+                        criterion = [criterion[0], "=", []];
```
Comment (Polymorphe57) [inline addons/web/static/tests/helpers/mock_server.js:1657]: why not in?

File: addons/web/static/tests/helpers/mock_server.js [hunk 6]
```diff
@@ -771,14 +770,13 @@ export class MockServer {
...
-            let [fieldName, aggregateFunction] = groupByField.split(":");
-            aggregateFunction = aggregateFunction || "month";
-            if (fields[fieldName].type === "date") {
-                // what about datetime?
-                if (!val) {
-                    return false;
-                }
-                const date = DateTime.fromISO(val);
+            if (!val) {
+                return false;
```
Comment (Polymorphe57) [inline addons/web/static/tests/helpers/mock_server.js:774]: does not work for type "integer" for example

File: addons/web/static/tests/helpers/mock_server.js [hunk 7]
```diff
@@ -790,101 +788,174 @@ export class MockServer {
...
+                        for (const groupValue of recordGroupValues) {
+                            acc.push({ ...groupValue, ...newGroup });
+                        }
+                    }
+                    return acc;
+                }, []);
+            }
+            for (const groupValue of recordGroupValues) {
+                const valueKey = JSON.stringify(groupValue);
+                groups[valueKey] = (groups[valueKey] || []).concat([r]);
```
Comment (Polymorphe57) [inline addons/web/static/tests/helpers/mock_server.js:846]: .push(r)

File: addons/web/static/tests/helpers/mock_server.js (hunk 5 above)
Comment (brboi) [reply addons/web/static/tests/helpers/mock_server.js:1657]: https://github.com/odoo/odoo/blob/94500b18365674e8ca8be68abee1918330efb553/addons/web/static/src/core/domain.js#L214

```js
[].some(a=>[].includes(a)) // returns false
```

and for the legacy mock server, it uses [_.intersection](http://underscorejs.org/#intersection)
https://github.com/odoo/odoo/blob/94500b18365674e8ca8be68abee1918330efb553/addons/web/static/src/legacy/js/core/domain.js#L106

File: addons/web/static/tests/helpers/mock_server.js (hunk 7 above)
Comment (brboi) [reply addons/web/static/tests/helpers/mock_server.js:846]: is that what you suggest ?
```suggestion
                groups[valueKey] = groups[valueKey] || [];
                groups[valueKey].push(r);
```

**Comment by Polymorphe57** [comment]
add some mockServer tests
also modif of new mockServer are not used right now

File: addons/web/static/tests/helpers/mock_server.js (hunk 7 above)
Comment (Polymorphe57) [reply addons/web/static/tests/helpers/mock_server.js:846]: for example but on same line is good too

File: addons/web/static/tests/helpers/mock_server.js (hunk 5 above)
Comment (Polymorphe57) [reply addons/web/static/tests/helpers/mock_server.js:1657]: [(field_name, "=", false)] won't work for many2many, in domain.js one returns fieldValue === value;

File: addons/web/static/tests/helpers/mock_server.js (hunk 7 above)
Comment (brboi) [reply addons/web/static/tests/helpers/mock_server.js:846]: I don´t know how to write this with Array.push in one line :-)

Array.push returns the new length of the array
```js
array = (array || []).push(42);
typeof array // = number
```

File: addons/web/static/tests/helpers/mock_server.js (hunk 5 above)
Comment (brboi) [reply addons/web/static/tests/helpers/mock_server.js:1657]: that is why the criterion[2] part is not used here ;-)

File: addons/web/static/tests/helpers/mock_server.js (hunk 6 above)
Comment (brboi) [reply addons/web/static/tests/helpers/mock_server.js:774]: true !

**Comment by Polymorphe57** [comment]
robodoo rebase-ff

**Comment by Polymorphe57** [comment]
robodoo r+

**Pull request closed without merging**